File: arrays/560_subarray_sum_equals_k.py
Brute Force

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0
        for i in range(len(nums)):
            curSum = 0
            for j in range(i, len(nums)):
                curSum += nums[j]
                if curSum == k:
                    res += 1

        return res

Prefix + Hash Map

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefixSums = {0 : 1}
        res, curSum = 0, 0

        for n in nums:
            curSum += n
            diff = curSum - k

            if diff in prefixSums:
                res += prefixSums[diff]
            prefixSums[curSum] = 1 + prefixSums.get(curSum, 0)

        return res

File: arrays/560_subarray_sum_equals_k/560_subarray_sum_equals_k.java
// Approach 1: Brute Force
// Time complexity: O(n²) 
// Space complexity: O(1)

class Solution {
    public int subarraySum(int[] nums, int k) {
        int res = 0;
        for (int i = 0; i < nums.length; i++) {
            int sum = 0;
            for (int j = i; j < nums.length; j++) {
                sum += nums[j];
                if (sum == k) {
                    res += 1;
                }
            }
        }
        return res;
    }
}

// Approach 2: Prefix + Hash Map
// Time complexity: O(n) 
// Space complexity: O(n) for hash map

class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> prefixSums = new HashMap<>();
        prefixSums.put(0, 1);
        int res = 0, curSum = 0;
        
        for (int n : nums) {
            curSum += n;
            int diff = curSum - k;

            res += prefixSums.getOrDefault(diff, 0);
            prefixSums.put(curSum, 1 + prefixSums.getOrDefault(curSum, 0));
        }
        
        return res;
    }
}
